changelog.py

```diff
@@ -0,0 +1,168 @@
+]
+
+
+class Changelog:
+    """A Changelog produces release notes for a single release."""
+
+    def __init__(self, repo):
+        self._repo = repo
+        self._template = Template(DEFAULT_TEMPLATE, trim_blocks=True)
+        self._ignore = set(self._slug(s) for s in DEFAULT_IGNORE)
+
+    def _slug(self, s: str) -> str:
+        """Return a version of the string that's easy to compare."""
+        return re.sub(r"[\s_-]", "", s.casefold())
+
+    def _previous_release(self, version: str):
+        """Get the release previous to the current one (according to SemVer)."""
+        cur_ver = semver.parse_version_info(version[1:])
+        prev_ver = semver.parse_version_info("0.0.0")
+        prev_rel = None
+        for tag in self._repo.tags.list(all=True):
+            if not tag.name.startswith("v"):
+                continue
+            try:
+                ver = semver.parse_version_info(tag.name[1:])
+            except ValueError:
+                continue
+            if ver.prerelease or ver.build:
+                continue
+            # Get the highest version that is not greater than the current one.
+            # That means if we're creating a backport v1.1, an already existing v2.0,
+            # despite being newer than v1.0, will not be selected.
+            if ver < cur_ver and ver > prev_ver:
+                prev_rel = tag
+                prev_ver = ver
+        return prev_rel
+
+    def _issues(self, start: datetime, end: datetime):
+        """Collect issues that were closed in the interval."""
+        issues = []
+        for x in self._repo.issues.list(state="closed", updated_after=start, all=True):
+            # If a previous release's last commit closed an issue, then that issue
+            # should be included in the previous release's changelog and not this one.
+            # The interval includes the endpoint for this same reason.
+            closed_at = parser.parse(x.closed_at)
+            if closed_at <= start or closed_at > end:
+                continue
+            if self._ignore.intersection(self._slug(label) for label in x.labels):
+                continue
+            else:
+                issues.append(x)
+        issues.reverse()  # Sort in chronological order.
+        return issues
+
+    def _merge_requests(self, start: datetime, end: datetime):
+        """Collect merge requests in the interval."""
+        merge_requests = []
+        for x in self._repo.mergerequests.list(state="merged", updated_after=start, all=True):
+            merged_at = parser.parse(x.merged_at)
+            if merged_at <= start or merged_at > end:
+                continue
+            if self._ignore.intersection(self._slug(label) for label in x.labels):
+                continue
+            else:
+                merge_requests.append(x)
+        merge_requests.reverse()  # Sort in chronological order.
+        return merge_requests
+
+    def _format_user(self, user) -> Dict[str, object]:
```

Comment:
If a function has a chance of returning none you can use the type hint `Optional`, such as `Optional[Dict[str, object]]`, `Optional[X]` is equivalent to `Union[X, None]`